src/resolver.rs

```diff
@@ -368,46 +370,123 @@ fn builtin_criteria() -> SortedMap<CriteriaName, CriteriaEntry> {
             },
         ),
     ]
-    .into_iter()
-    .collect()
 }
 
 impl CriteriaMapper {
-    pub fn new(criteria: &SortedMap<CriteriaName, CriteriaEntry>) -> CriteriaMapper {
-        let builtins = builtin_criteria();
-        let list = criteria
+    pub fn new(
+        criteria: &SortedMap<CriteriaName, CriteriaEntry>,
+        imports: &ImportsFile,
+        mappings: &SortedMap<ImportName, RemoteImport>,
+    ) -> CriteriaMapper {
+        // First, build a list of all our criteria
+        let locals = criteria
             .iter()
-            .chain(builtins.iter())
-            .map(|(k, v)| (k.clone(), v.clone()))
+            .map(|(k, v)| (CriteriaNamespace::Local, k.clone(), v.clone()));
+        let builtins = builtin_criteria();
+        let foreigns = imports.audits.iter().flat_map(|(import, audit_file)| {
+            audit_file.criteria.iter().map(|(k, v)| {
+                (
+                    CriteriaNamespace::Foreign(import.clone()),
+                    k.clone(),
+                    v.clone(),
+                )
+            })
+        });
+        let list = builtins
+            .into_iter()
+            .chain(locals)
+            .chain(foreigns)
             .collect::<Vec<_>>();
-        let index: FastMap<CriteriaName, usize> = list
-            .iter()
-            .enumerate()
-            .map(|(idx, v)| (v.0.clone(), idx))
-            .collect();
+        let num_criteria = list.len();
 
-        let mut implied_criteria = Vec::with_capacity(list.len());
-        for (idx, (_name, entry)) in list.iter().enumerate() {
-            // Precompute implied criteria (doing it later is genuinely a typesystem headache)
-            let mut implied = CriteriaSet::none(list.len());
-            implied.set_criteria(idx);
-            recursive_implies(&mut implied, &entry.implies, &index, &list);
+        // Now construct an index over this list, that lets us map
+        // (CriteriaNamespace, CriteriaName) => CriteriaIdx
+        let mut index = FastMap::<CriteriaNamespace, FastMap<CriteriaName, usize>>::new();
+        let all_import_names = mappings.iter().map(|(import_name, _)| import_name);
+
+        // Add all the natural entries
+        for (idx, (namespace, criteria_name, _entry)) in list.iter().enumerate() {
+            let prev = index
+                .entry(namespace.clone())
+                .or_default()
+                .insert(criteria_name.clone(), idx);
+            assert!(prev.is_none(), "criteria name was multiply defined???");
+        }
+
+        // Hack the builtins into every foreign namespace. Because they all use
+        // the same CriteriaIdx, we can now forget that builtins are special and
+        // just naturally look them up in any namespace without issue.
```

Comment:
Hmm this might not work quite right when using namespace as AuditSource when e.g. handling blames? I think it works out but needs consideration.